Add CORS setting and addNote audio downloading support

sp3ctum opened the pull request:
# Adds a new setting called `cors_host`
This setting, when set, enables CORS (cross origin resource sharing) headers in the API. These allow browser access from the configured site. By default, browsers are denied access, and only when CORS headers are applied is the access given.

The new setting can be configured in the settings:

<img width="450" alt="image" src="https://user-images.githubusercontent.com/300791/227700747-2f86b55e-1f08-4dff-bc69-751c100d76e8.png">

Related issue: https://github.com/KamWithK/AnkiconnectAndroid/issues/13

# Adds support for the addNote api to download audio files from given URLs
An example of such a request can be seen here: https://github.com/KamWithK/AnkiconnectAndroid/issues/17

Aquafina-water-bottle commented on app/src/main/java/com/kamwithk/ankiconnectandroid/routing/RouteHandler.java:
```diff
@@ -52,10 +58,22 @@ public NanoHTTPD.Response get(RouterNanoHTTPD.UriResource uriResource, Map<Strin
 
-        NanoHTTPD.Response rep = apiHandler.chooseAPI(files.get("postData"), session.getParameters());
+        NanoHTTPD.Response rep;
+        if (files.get("postData") == null) {
+            rep = newFixedLengthResponse("Ankiconnect Android is running.");
+        } else {
+            rep = apiHandler.chooseAPI(files.get("postData"), session.getParameters());
+        }
```
Can you make sure this doesn't inadvertently affect Forvo audio? If memory serves me correct, I tried to add this in a different PR, but removed it because the Forvo audio was no longer getting fetched.

P.S. Thanks for the PR! I'll try to make some time to review it properly sometime in the next few days.

sp3ctum replied:
Sure! I'll set up my yomichan according to the instructions in the readme.

No rush with the PR, take your time!

sp3ctum replied:
You were right - forvo audio was broken by this change. I think I was able to fix it, though: see 87c2ecd7c034e66734915b2ea3d76c01da671128

In yomichan, I can now select all these audios and hear the word being read by different people
![image](https://user-images.githubusercontent.com/300791/227704369-458c5f78-cc32-4ae0-a198-61d2605c48c8.png)

Aquafina-water-bottle replied:
Looks good, thanks for the change! I'll temporarily keep this open as a self reminder to verify that it works on my end.

sp3ctum replied:
Just noticed I had accidentally broken my cors support with the fix. :)
Pushed some more changes - now forvo works locally and the cors headers are also applied correctly.

Aquafina-water-bottle replied:
Seems like everything here works, thanks!

KamWithK commented:
Thanks for the contribution, I'll take a look when I get some time!

Aquafina-water-bottle commented on app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java:
```diff
@@ -78,8 +80,22 @@ public Long addNote(final Map<String, String> data, String deck_name, String mod
+        audioFile.setData(data);
+
+        String filePath = storeMediaFile(audioFile);
+        return filePath;
```
I believe these changes should be completely moved to `MediaAPI`. My understanding of `IntegratedAPI` is that this class is only used when:
- More than one API is being used (e.g. `addNote`)
- The Android `Context` is needed (e.g. `guiBrowse`)

Aquafina-water-bottle commented on app/src/main/java/com/kamwithk/ankiconnectandroid/routing/AnkiAPIRouting.java:
```diff
@@ -159,19 +177,41 @@ private String canAddNotes(JsonObject raw_json) {
+        audioFile.setFilename(Parser.getMediaFilename(raw_json));
+        audioFile.setData(Parser.getMediaData(raw_json));
+
+        return Parser.gson.toJson(integratedAPI.storeMediaFile(audioFile));
```
Fortunately, these changes still work on images (I tested this on a Yomichan dictionary with images, and it exported perfectly fine). However, changing this is a misnomer, because the media can be images or audio (and theoretically, anything other byte array). I think it would be safe to undo the logic, or rename the `AudioFile` class to something like `BinaryFile`.

Aquafina-water-bottle commented on app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/MediaAPI.java:
```diff
@@ -25,9 +29,14 @@ public MediaAPI(Context context) {
     @SuppressLint("SetWorldReadable")
     public String storeMediaFile(String filename, byte[] data) throws IOException {
-        File file = new File(context.getCacheDir(), Uri.parse(filename).getLastPathSegment());
+        String lastPathSegment = Uri.parse(filename).getLastPathSegment();
+        lastPathSegment = lastPathSegment == null ? filename : lastPathSegment;
+        File file = new File(context.getCacheDir(), lastPathSegment);
```
Just curious about this: did adding this logic solve something on your end?

Aquafina-water-bottle reviewed (changes requested):
Overall, everything looks good! Especially appreciate the documentation in the code. There were just a few nitpicks and questions.

Note that I didn't actually test the main parts of this PR (downloading audio from `addNote` and the CORS additions), but I tried running it through the standard Yomichan setup, and tested everything else I think it could've affected, and everything there seems to work fine for me. (We might want to eventually add some tests to this, but that's outside the scope of this PR.)

P.S. Thank you for getting the `Ankiconnect Android is running.` message working! It'll be very useful to debug people's setups, and once this gets merged, I'll make sure to add it somewhere in the README.

Aquafina-water-bottle commented:
The changes look good, thanks! One last thing, as mentioned in #31, we're planning on quickly creating a `docs/api.md` file, where edge cases such `addNote` not supporting downloaded files can be properly documented. I'm thinking it would be best to block merging this PR until that file gets added to master, and until you can add this call into the file.

KamWithK commented:
In agreement with @Aquafina-water-bottle about blocking this until docs are added

sp3ctum commented:
Sure! I subscribed to that PR and will add documentation after it gets merged.

KamWithK commented:
Now that the documentation PR has been merged, add some docs and then once you're ready we can take another look at this PR @sp3ctum

sp3ctum commented:
Just added (a little) documentation and squashed the commits for a cleaner history. I think this is ready for another review round whenever you have the time.

Aquafina-water-bottle commented:
Thanks for the documentation, everything looks good to me!

KamWithK commented:
Looks good thanks @sp3ctum @Aquafina-water-bottle
